Add support for additional observation data in processing pipeline

xincjin-NOAA opened the pull request.

azadeh-gh commented:
Thanks for the great work structuring the feature extraction by obs_type and instrument_name. To fully the updated GNN model and ocelot_loss (which uses instrument_ids and padded feature vectors), a few key updates are needed:
1.  Determine max feature dimension across all instruments:  Before looping over bins, calculate the maximum number of features across all instruments (after metadata + lat/lon + obs features).

2. Pad input and target features: For each instrument in each bin, pad input_features_final and target_features_final to the max length using zeros. This ensures feature vectors from all instruments can be stacked together.

3. Assign instrument_id: Define a consistent mapping (e.g., {"atms": 0, "surface_pressure": 1, "radiosonde": 2, ...}). Add a field instrument_ids = np.full((num_rows,), instrument_id) and attach it to each instrument’s entry.

4. Flatten across instruments: After extracting/padding features and metadata per instrument, concatenate everything at the bin level:
````python
data_summary[bin_name]["input_features_final"] = np.vstack([...])
data_summary[bin_name]["instrument_ids"] = np.concatenate([...])
````
Do the same for targets and metadata (target_features_final, target_metadata, etc.).

**Why this is needed:** 
Our model expects a single matrix per bin, with uniform input feature size and a parallel instrument_ids tensor for weighting loss.
Without padding and flattening, we won’t be able to batch across instruments.


What the model expects (after merging/padding):

Let’s say:
ATMS has 3 observations with 3 channels (c1, c2, c3)
Surface Pressure has 2 observations with 1 variable (p)
We pad Surface Pressure to have the same number of columns as ATMS (fill with 0 for unused channels), and assign instrument IDs like so:

![image](https://github.com/user-attachments/assets/74f35bb2-cd57-4129-8728-10548da022c9)

- Each row is an observation, with latitude, longitude, features, and an instrument_id.
- ATMS (instrument_id = 0) has values in all 3 channels (c1, c2, c3).
- Surface Pressure (instrument_id = 1) uses only c1 (with normalized pressure), and pads c2, c3 with zeros.
- The instrument_id column lets the model apply the correct weights or logic per instrument.


**Key Ideas:** 

- All observations are merged into one matrix (x), padded as needed.
- instrument_id tells the model what the features actually mean.
- Loss is computed per instrument using their own channel weights and valid channel count.
- This allows us to train on all instruments together, even if they have different feature sizes.

azadeh-gh commented:
Please take a look at the [feature/loss branch](https://github.com/NOAA-EMC/ocelot/tree/feature/loss) — it includes:

- A new `ocelot_loss` that supports instrument-wise and channel-wise weighting
- Instrument ID assignment + zero-padding for unequal channel sizes
- `weights_config.yaml`
- Modified versions of `gnn_model.py`, `gnn_datamodule.py`, and `train_gnn.py` to work with the updated structure

The training setup expects a flattened and padded structure per bin (merged across instruments) to be passed to the model. Right now, process_timeseries.py returns a nested structure, so it will need a final flattening step to align with this.

Let me know if you’d like help with that part!

azadeh-gh commented:
> @azadeh-gh I'm planning to work on the "Flatten across instruments" feature and the related items you mentioned. Should I implement these outside of extract_features, or would you prefer they be integrated within it?

@xincjin-NOAA Thanks for asking! I recommend implementing the "flatten across instruments" step outside of extract_features().

xincjin-NOAA commented:
Flattened data_summary by combining obs_type and instrument_name so it's ready for the next step in the pipeline
-  Feature Padding: For each feature type (input, target, metadata):
    - Pads shorter features with zeros to match maximum dimensions
- Data Preservation: 
  - Copies non-tensor data directly: scalar parameters (min/max) and geographical coordinates (lat/lon)

xincjin-NOAA commented:
Refactored the flatten code and tested the new implementation.

- Key improvements include: Extracted the padding logic into a reusable helper function, Centralized all tensor collections using a single feature_lists dictionary, Simplified the processing loop by iterating directly over flattened_variables.
- These changes result in more maintainable code with reduced duplication, making it easier to extend with additional features or metadata types.

xincjin-NOAA commented on gnn_model/process_timeseries.py:
```diff
@@ -124,83 +173,206 @@ def extract_features(z, data_summary):
+                data_summary_bin["target_metadata"] = torch.tensor(target_metadata, dtype=torch.float32)
+                # Store min/max values for later unnormalization
+                data_summary_bin["target_scaler_min"] = target_scaler.data_min_
+                data_summary_bin["target_scaler_max"] = target_scaler.data_max_
```
corrected.
not tested yet at that time.

Thanks!

xincjin-NOAA commented:
Merged main into this branch

Validated after merge with some refactoring accordingly. 

Finished end2end run.

xincjin-NOAA commented:
@azadeh-gh Refactored, thanks.

mlaurako commented on gnn_model/process_timeseries.py:
```diff
@@ -6,63 +6,100 @@
 
 @timing_resource_decorator
-def organize_bins_times(z, start_date, end_date, selected_satids):
+def organize_bins_times(z_dict, start_date, end_date, observation_config, bin_size='12h'):
```
I tested with bin_size='6h'. The code unfortunately ran into this error, happened during encoding. Does this happen to you?

`RuntimeError: mat1 and mat2 shapes cannot be multiplied (55932x10 and 33x32)`

xincjin-NOAA replied:
Could you provide some more details? The codes do stopped somewhere when I run end2end with '6h'.  However, I haven't find the place that failed. 

Thanks!

mlaurako replied:
Sure thing. Here's the error file </scratch1/NCEPDEV/da/Mu-Chieh.Ko/ocelot/Xin_PR/ocelot/gnn_model/gnn_train_12226805.err> if you would like to take a look.

xincjin-NOAA commented:
moved "height' to metadata.

xincjin-NOAA commented on gnn_model/train_gnn.py:
```diff
@@ -22,17 +22,43 @@ def main():
+        "conventional": {
+            # "radiosonde": ,
+            "pressure": {
+                "features": ["height", "stationPressure"]
```
moved the height to the metadata.
